Reflection: Fix crash in single-payload enum layout if payload type c…

slavapestov opened the pull request:
- Description: The reflection library would dereference a null pointer if it encountered an `Optional` of a type for which we do not have reflection metadata (possibly because we did a release build, or if IRGen neglects to emit the info due to a bug).

- Scope of the issue: Affects anyone doing memory graph debugging in Instruments when the above scenario occurs.

- Risk: Low, the existing logic mostly already handles missing type info, but in this case we forgot to do a null check.

- Tested: Existing tests pass, and there's a new test for the scenario that used to crash.

- Reviewed by: David Farler (pending)

- Radar: rdar://problem/27906876

I will continue investigating the issue (in particular, I'm wondering what the root cause of the missing type info is, since in theory we should always emit everything that we need). But even if we fix the original problem, we need this fix anyway, for general robustness.

slavapestov commented:
@tkremenek This is still WIP because I'd like to come up with a test case, but the fix itself is sound I think. I'll post an update later tonight.

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
@bitjammer Do you mind reviewing this for 3.0? Sorry for the short notice.

I will add a test case that reproduces this specific issue shortly (we need to compile two .swift files, one without -enable-reflection-metadata). Still figuring out why the project in the radar has missing type info though.

bitjammer commented:
@slavapestov The change seems reasonable to me, although we should really try to get a test in for this. To reproduce with the example project, you have to build with the Release configuration (which is weird).

slavapestov commented:
So were you able to reproduce it? Can you tell me how to change the build configuration?

I have a test now, btw. Just making sure it passes first :)

bitjammer commented:
Just click the scheme next to run/stop, `Edit Scheme...` > `Info` > `Build Configuration`. At the command line, I think you can `xcodebuild -configuration Release` too if you want.

slavapestov commented:
Thanks @bitjammer!

slavapestov commented:
Ok, the graph debugger didn't come up that time, where do I see the crash log?

slavapestov commented:
Yeah, it looks like we just have an `Optional` of a type that did not have reflection info emitted, possibly because it's a release build.

bitjammer commented:
Were you able to get the mangled name of the offending type? Just looking at the code, it might be one of the Objective-C types like NSWindow or NSButton? I assume `weak` isn't the problem here ... or is it?

slavapestov commented:
No, I wasn't. We don't pass -enable-reflection-metadata in release builds, do we? Or is that not the problem?

slavapestov commented:
I found another place where a null return from getTypeInfo() triggers an assertion -- will push a better fix shortly.

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
@bitjammer I added a test, and another fix for null type info. We were also using null to denote an invalid recursive value type, which is wrong since we end up with null when we reference a missing type. I changed the code to use a separate set to track the TypeRefs that are in the process of being lowered. Do you mind quickly reviewing again?

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
@tkremenek This is now ready to go in, pending David reviewing the second version and CI passing.

bitjammer commented:
LGTM.

slavapestov commented:
Thanks @bitjammer
